pkg/asset/machines/master.go

```diff
@@ -114,9 +138,110 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 	default:
 		return fmt.Errorf("invalid Platform")
 	}
+
+	machines := []interface{}{}
+	if len(m.Machines) == 0 {
+		for _, machine := range m.MachinesDeprecated {
+			machines = append(machines, machine)
+		}
+	} else {
+		for _, machine := range m.Machines {
+			machines = append(machines, machine)
+		}
+	}
+
+	for i, machine := range machines {
+		data, err := yaml.Marshal(machine)
+		if err != nil {
+			return errors.Wrapf(err, "marshal master %d", i)
+		}
+
+		m.FileList = append(m.FileList, &asset.File{
+			Filename: filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, strconv.Itoa(i))),
+			Data:     data,
+		})
+	}
+
+	asset.SortFiles(m.FileList)
+
 	return nil
 }
 
+// Files returns the files generated by the asset.
+func (m *Master) Files() []*asset.File {
+	return m.FileList
+}
+
+// Load reads the asset files from disk.
+func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
+	file, err := f.FetchByName(filepath.Join(directory, MasterUserDataFileName))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	m.FileList = []*asset.File{file}
+
+	m.Machines = []machineapi.Machine{}
+	m.MachinesDeprecated = []clusterapi.Machine{}
+	fileList, err := f.FetchByPattern(filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, "*")))
+	if err != nil {
+		return true, err
+	}
+	m.FileList = append(m.FileList, fileList...)
+	for _, file := range fileList {
+		machine := &machineapi.Machine{}
```

Comment:
Are we comfortable that a `clusterapi.Machine` will successfully unmarshal into a `machineapi.Machine`? The safer approach would be to unmarshal into a `map[string]interface{}` first and inspect the `apiVersion` field.

Reply:
Or into a `runtime.TypeMeta` like you do to get the provider type.

Reply:
> The safer approach would be to unmarshal into a `map[string]interface{}` first and inspect the `apiVersion` field.

Currently:

```console
$ head -n1 wking/openshift/99_openshift-cluster-api_master-machines-0.yaml 
apiVersion: 5
trking@trking ~/.local/lib/go/src/github.com/openshift/installer $ openshift-install --dir=wking create ignition-configs
FATAL failed to fetch Bootstrap Ignition Config: failed to load asset "Master Machines": unmarshal "openshift/99_openshift-cluster-api_master-machines-0.yaml": error unmarshaling JSON: json: cannot unmarshal number into Go struct field Machine.apiVersion of type string 
```

I think that's useful validation (and it covers the rest of the structure, except the opaque-at-this-stage provider spec, which we unmarshal a few lines down).  Using `interface{}`-style marshalling would give us the opportunity to customise error messages, but I think the above error is clear enough that I'm not excited about sinking time into the interface approach ;).  Thoughts?

Reply:
I am confused about what you are showing me here and what the conclusion is. I was expecting that `apiVersion` was either `machine.openshift.io/v1beta1` or `cluster.k8s.io/v1alpha1`, not `5`. However, you are comfortable that an openstack master machine file can subsequently be unmarshaled into a `machineapi.Machine` enough to determine the provider type, then I am satisfied.

Reply:
> I was expecting that `apiVersion` was either `machine.openshift.io/v1beta1` or `cluster.k8s.io/v1alpha1`, not `5`. 

I was trying to show what errors looked like when unmarshal failed (to cover your "safer" concerns).

> However, if you are comfortable that an openstack master machine file can subsequently be unmarshaled into a machineapi.Machine enough to determine the provider type, then I am satisfied.

I'm don't have an OpenStack account to test with, so I get:

```console
$ cat wking/install-config.yaml
apiVersion: v1beta2
baseDomain: devcluster.openshift.com
metadata:
  name: wking
platform:
  openstack: {}
pullSecret: ...redacted...
$ openshift-install --dir=wking create manifests
FATAL failed to fetch Master Machines: failed to load asset "Install Config": invalid "install-config.yaml" file: platform.openstack.cloud: Internal error: could not retrieve valid clouds 
```

Maybe @flaper87 or @tomassedovic can give this PR a spin on OpenStack to let us know if:

```console
$ openshift-install --dir=wking create manifests
$ openshift-install --dir=wking create ignition-configs
```

works?  In the mean time, I'll see if I can hack around the missing stuff to get my installer to generate OpenStack Machine configs.

Reply:
I was able to generate the following `99_openshift-cluster-api_master-machines-0.yaml` file for an OpenStack machine. The installer was able to load it successfully. If we are comfortable this the deprecated machines are temporary and we will not keep this code through as `clusterapi` and `machineapi` drift apart, then I am fine with the code as is.

```
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  creationTimestamp: null
  labels:
    sigs.k8s.io/cluster-api-cluster: mstaeble
    sigs.k8s.io/cluster-api-machine-role: master
    sigs.k8s.io/cluster-api-machine-type: master
  name: mstaeble-master-0
  namespace: openshift-cluster-api
spec:
  metadata:
    creationTimestamp: null
  providerSpec:
    value:
      apiVersion: openstackproviderconfig.k8s.io/v1alpha1
      cloudName: openstack
      cloudsSecret:
        name: openstack-credentials
        namespace: kube-system
      flavor: myFlavor
      image: rhcos
      kind: OpenstackProviderSpec
      networks:
      - filter:
          tags: openshiftClusterID=653b7d0d-c273-4007-a39a-155ab2ca5e9d
      root_volume:
        volumeType: ""
      securityGroups:
      - master
      userDataSecret:
        name: master-user-data
  versions:
    kubelet: ""
status: {}
```

Reply:
> In the mean time, I'll see if I can hack around the missing stuff...

With

```diff
diff --git a/pkg/asset/installconfig/platformcredscheck.go b/pkg/asset/installconfig/platformcredscheck.go
index 671ae14..8e78456 100644
--- a/pkg/asset/installconfig/platformcredscheck.go
+++ b/pkg/asset/installconfig/platformcredscheck.go
@@ -30,2 +30,3 @@ func (a *PlatformCredsCheck) Dependencies() []asset.Asset {
 func (a *PlatformCredsCheck) Generate(dependencies asset.Parents) error {
+	return nil
 	ic := &InstallConfig{}
diff --git a/pkg/asset/manifests/openshift.go b/pkg/asset/manifests/openshift.go
index c63a9d7..f67d8dd 100644
--- a/pkg/asset/manifests/openshift.go
+++ b/pkg/asset/manifests/openshift.go
@@ -15,5 +15,5 @@ import (
 	// https://github.com/openshift/installer/pull/854
-	goyaml "gopkg.in/yaml.v2"
+//	goyaml "gopkg.in/yaml.v2"
 
-	"github.com/gophercloud/utils/openstack/clientconfig"
+//	"github.com/gophercloud/utils/openstack/clientconfig"
 	"github.com/openshift/installer/pkg/asset"
@@ -21,3 +21,3 @@ import (
 	"github.com/openshift/installer/pkg/asset/machines"
-	osmachine "github.com/openshift/installer/pkg/asset/machines/openstack"
+//	osmachine "github.com/openshift/installer/pkg/asset/machines/openstack"
 	"github.com/openshift/installer/pkg/asset/password"
@@ -85,2 +85,3 @@ func (o *Openshift) Generate(dependencies asset.Parents) error {
 	case "openstack":
+/*
 		opts := new(clientconfig.ClientOpts)
@@ -107,2 +108,3 @@ func (o *Openshift) Generate(dependencies asset.Parents) error {
 		}
+*/
 	}
diff --git a/pkg/types/openstack/validation/platform.go b/pkg/types/openstack/validation/platform.go
index 1217978..4b5a662 100644
--- a/pkg/types/openstack/validation/platform.go
+++ b/pkg/types/openstack/validation/platform.go
@@ -13,2 +13,3 @@ func ValidatePlatform(p *openstack.Platform, fldPath *field.Path, fetcher ValidV
 	allErrs := field.ErrorList{}
+	return allErrs
 	validClouds, err := fetcher.GetCloudNames()

```

and [my earlier `install-config.yaml`][1], I can generate `openshift/99_openshift-cluster-api_master-machines-0.yaml` like:

```yaml
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  creationTimestamp: null
  labels:
    sigs.k8s.io/cluster-api-cluster: wking
    sigs.k8s.io/cluster-api-machine-role: master
    sigs.k8s.io/cluster-api-machine-type: master
  name: wking-master-0
  namespace: openshift-cluster-api
spec:
  metadata:
    creationTimestamp: null
  providerSpec:
    value:
      apiVersion: openstackproviderconfig.k8s.io/v1alpha1
      cloudName: openstack
      cloudsSecret:
        name: openstack-credentials
        namespace: kube-system
      flavor: ""
      image: rhcos
      kind: OpenstackProviderSpec
      networks:
      - filter:
          tags: openshiftClusterID=6b7a511b-e595-4077-95fa-11cd058f70e0
      root_volume:
        volumeType: ""
      securityGroups:
      - master
      userDataSecret:
        name: master-user-data
  versions:
    kubelet: ""
status: {}
```

A subsequent `openshift-install --dir=wking create ignition-configs` sucks them back in without crashing, so I think we're good here.  @staebler, did you want me to check anything else in this space?

[1]: https://github.com/openshift/installer/pull/1211#discussion_r255217192

Reply:
Nope. I am fine with this if we are sure that `MachineDeprecated` will expire soon.